test/upon_error_test.cpp

```diff
@@ -170,6 +170,16 @@ TEST(UponError, ManyErrorSender) {
   >);
 }
 
+TEST(UponError, ManyErrorSenderAllReturnInt) {
+  auto s = many_error_sender{} | upon_error([](auto) -> int {
+    return 0;
+  });
+  static_assert(std::is_same_v<
+    decltype(s)::value_types<std::variant, std::tuple>,
+    std::variant<std::tuple<double>, std::tuple<int>>
```

Comment:
My main goal of this PR was to fix the output sender from `upon_error` to not contain duplicate values in `value_types`. I chose to change the behavior of `concat_type_lists_unique_t`, but I'm happy to take another approach (such as directly changing upon_error and possibly other algos that might have a similar problem).